sumpy/tools.py

```diff
@@ -931,6 +932,100 @@ def to_complex_dtype(dtype):
     except KeyError:
         raise RuntimeError(f"Unknown dtype: {dtype}")
 
+
+@memoize
+def get_real_part_kernel(context, dtype):
+    from pyopencl.elementwise import ElementwiseKernel
+    pyopencl_name = "float" if dtype == np.float32 else "double"
+    return ElementwiseKernel(context,
+        f"c{pyopencl_name}_t *x, {pyopencl_name} *z",
+        f"z[i] = c{pyopencl_name}_real(x[i])",
+        "real_part",
+        preamble="#include <pyopencl-complex.h>")
+
+
+ProfileGetter = namedtuple("ProfileGetter", "start, end")
+
+
+class AggregateProfilingEvent:
+    """An object to hold a list of events and provides compatibility
+    with some of the functionality of :class:`pyopencl.Event`.
+    Assumes that the last event waits on all of the previous events.
+    """
+    def __init__(self, events):
+        self.events = events[:]
+        if isinstance(events[-1], cl.Event):
+            self.native_event = events[-1]
+        else:
+            self.native_event = events[-1].native_event
+
+    @property
+    def profile(self):
+        total = sum(evt.profile.end - evt.profile.start for evt in self.events)
+        end = self.native_event.profile.end
+        return ProfileGetter(start=end - total, end=end)
+
+    def wait(self):
+        return self.native_event.wait()
+
+
+class MarkerBasedProfilingEvent:
+    """An object to hold two marker events and provides compatibility
+    with some of the functionality of :class:`pyopencl.Event`.
+    """
+    def __init__(self, *, end_event, start_event):
+        self.native_event = end_event
+        self.start_event = start_event
+
+    @property
+    def profile(self):
+        return ProfileGetter(start=self.start_event.profile.start,
+                             end=self.native_event.profile.end)
+
+    def wait(self):
+        return self.native_event.wait()
+
+
+@memoize(use_kwargs=True)
+def cached_vkfft_app(*args, **kwargs):
+    from pyvkfft.opencl import VkFFTApp
+    return VkFFTApp(*args, **kwargs)
+
+
+def run_opencl_fft(queue, input_vec, output_vec=None, inverse=False, wait_for=None):
+    """Runs a FFT on input_vec and returns two opencl markers that indicate the
+    end and start of the operations carried out. Only supports in-order queues.
+    If output_vec is given, the FFT is run out-of-place and the input_vec
+    is overwritten iff it is an inverse FFT with the output_vec type is real.
+    If output_vec is not given, an in-place FFT on input_vec is run.
+    """
+    if queue.properties & cl.command_queue_properties.OUT_OF_ORDER_EXEC_MODE_ENABLE:
+        raise RuntimeError("VkFFT does not support out of order queues yet.")
+
+    if wait_for is None:
+        wait_for = []
+
+    start_evt = cl.enqueue_marker(queue, wait_for=wait_for[:])
+
+    input_dtype = input_vec.dtype.type
+    assert input_dtype in (np.float32, np.float64, np.complex64,
+                           np.complex128)
+
+    inplace = output_vec is None
```

Comment:
Yes, all of the operations are in-place. (I also removed out-of-place from this function because there were no uses of out-of-place)

I don't think it's dangerous. Any reason why you think it's dangerous?